src/index.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * @fileoverview This is an ES Module which exports `vizObserver`, see its function below.
+ *
+ * This uses a combination of IntersectionObserver and ResizeObserver to inform you when DOM
+ * nodes move around on a page (and also when they're resized). This might be useful if
+ * you're trying to track a complex HTML structure, perhaps one you don't control.
+ *
+ * It also works with target nodes inside Shadow DOM.
+ */
+
+/**
+ * This is an element's position relative to the page's documentElement.
+ */
+export interface Rect {
+  appear: boolean;
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+}
+
+export interface Options {
+  signal?: AbortSignal;
+}
+
+// const debug = false;
+const root = document.documentElement;
+
+const handlers = new WeakMap<Element, () => void>();
+
+const activeObservers = new Set<() => void>();
+handlers.set(root, () => activeObservers.forEach((handler) => handler()));
+
+const ro = new ResizeObserver((entries) => {
+  for (const entry of entries) {
+    handlers.get(entry.target)?.();
+  }
+});
+
+/**
+ * Sets up an observer that notifies on element resize or move.
+ *
+ * Returns a method to remove the observer (or accepts an AbortSignal). You should be sure to clean
+ * up outstanding observers even if you otherwise discard the element, otherwise you could leak
+ * memory.
+ */
+export default function vizObserver(element: Element, callback: (rect: Rect) => void) {
+  let io: IntersectionObserver | null = null;
+
+  // const viz = document.createElement('div');
+  // viz.className = 'viz';
+  // debug && document.body.append(viz);
+
+  const refresh = (threshold = 1.0) => {
+    io && io.disconnect();
+    io = null;
+
+    const rect = element.getBoundingClientRect();
+    const se = document.scrollingElement;
+    const top = rect.top + (se?.scrollTop ?? 0);
+    const left = rect.left + (se?.scrollLeft ?? 0);
+
+    if (!rect.width || !rect.height) {
+      callback({ appear: false, x: 0, y: 0, width: 0, height: 0 });
+      return; // Wait for the element to be resized to a sensible size.
+    }
+    callback({
+      appear: true,
+      x: left,
+      y: top,
+      width: rect.width,
+      height: rect.height,
+    });
+
+    // Calculate the exact position this element holds on the page.
+    const x = (v: number) => Math.floor(v);
```

Comment:
(would be compiled out anyway, make it clearer)

```suggestion
    const floor = (v: number) => Math.floor(v);
```